runtime/parachains/src/shared.rs

```diff
@@ -64,6 +71,40 @@ pub mod pallet {
 	#[pallet::getter(fn active_validator_keys)]
 	pub(super) type ActiveValidatorKeys<T: Config> = StorageValue<_, Vec<ValidatorId>, ValueQuery>;
 
+	/// All included blocks on the chain, mapped to the core index the candidate was assigned to
+	/// and the block number in this chain that should be reverted back to if the candidate is
+	/// disputed and determined to be invalid.
+	#[pallet::storage]
+	#[pallet::getter(fn included_candidates)]
+	pub(super) type IncludedCandidates<T: Config> = StorageDoubleMap<
+		_,
+		Twox64Concat,
+		SessionIndex,
+		Blake2_128Concat,
+		CandidateHash,
+		(T::BlockNumber, CoreIndex),
+	>;
+
+	/// The set of all past sessions that have yet to fully pruned. Sessions are added to this set
+	/// upon moving to new current session, and removed after all of their included candidates
+	/// have been removed. This set is tracked independently so that the cost of removing included
+	/// candidates can be amortized over multiple blocks rather than removing all candidates upon
+	/// transitioning to a new session.
+	#[pallet::storage]
+	#[pallet::getter(fn pruneable_sessions)]
+	pub(super) type PruneableSessions<T: Config> = StorageMap<_, Twox64Concat, SessionIndex, ()>;
```

Comment:
@thiolliere if this pattern is legit, it would be easy to abstract it behind a `StorageSet`.